devlog/2024-02-09-Intro-to-optimizing-packages-on-solus.md

```diff
@@ -0,0 +1,465 @@
+
+```
+$ hyperfine "dwebp ~/3.webp -o /dev/null"
+Benchmark 1: dwebp ~/3.webp -o /dev/null
+  Time (mean ± σ):     198.6 ms ±   2.3 ms    [User: 194.3 ms, System: 3.6 ms]
+  Range (min … max):   196.3 ms … 203.3 ms    14 runs
+$ sudo eopkg it zlib-2.1.5-28-1-x86_64.eopkg
+...
+$ $ hyperfine "dwebp ~/3.webp -o /dev/null"
+Benchmark 1: dwebp ~/3.webp -o /dev/null
+  Time (mean ± σ):      87.6 ms ±   0.7 ms    [User: 84.7 ms, System: 2.6 ms]
+  Range (min … max):    86.5 ms …  88.7 ms    33 runs
+```
+
+Well, hot diggity damn. Swapping out the zlib package for a more performant variant has instantly more than halved(!!) our decoding time.
+
+We need to find a more contained `libpng` benchmark here that removes the `libwebp` stuff to really confirm the findings here. After some sleuthing the `libpng` source repository has a [pngtopng.c](https://github.com/glennrp/libpng/blob/libpng16/contrib/examples/pngtopng.c) file we can compile to use the system libpng library.
+
+Changing the header from `#include "../../png.h"` to `#include <png.h>` then running `gcc -Ofast pngtopng.c -lpng16 -o pngtopng` to compile it, we have a libpng benchmark. We can reuse our test .png file from earlier. Ending up with: `./pngtopng ~/PNG_Test.png /dev/null` for our benchmark.
+
+| Library          | Time     |
+| ---------------- | -------- |
+| zlib (canonical) | 1.464 s  |
+| zlib-ng (compat) | 896.6 ms |
+
+Well. This is pretty much inline with our flawed `dwebp` benchmark from earlier. Swapping out zlib almost halves `libpng` decoding time.
+
+## Squeezing more from zlib-ng
+
+However, we're not done yet. We still have our compiler optimizations available to us to squeeze more performance from `zlib-ng`.
+
+| Optimization                          | Decode             | Size      |
+| ------------                          | ------------------ | --------- |
+| Baseline                              | 896.6 ms           | 141.00 KB |
+| Speed                                 | 883.6 ms           | 182.00 KB |
+| LTO                                   | 892.7 ms           | 133.00 KB |
+| PGO                                   | 894.6 ms           | 141.00 KB |
+| x86-64-v3                             | 892.5 ms           | 295.00 KB |
+| Speed + LTO                           | 882.6 ms           | 170.00 KB |
+| Speed + LTO + PGO + x86-64-v3         | 882.5 ms           | 250.00 KB |
+
+It looks like in this case the simple speed + LTO optimizations is the way to go. Speed gives the majority of the speedup but LTO helps bring back down the package size again. However, it's only a 1.5% improvement from baseline for this benchmark. We can always rebenchmark it later, testing zlib performance more directly instead of via libpng. It shows how good job the zlib-ng developers have done that it's so performant right out of the gate.
```

Comment:
```suggestion
It looks like in this case the simple speed + LTO optimizations is the way to go. Speed gives the majority of the speedup but LTO helps bring back down the package size again. However, it's only a 1.5% improvement from baseline for this benchmark. We can always rebenchmark it later, testing zlib performance more directly instead of via libpng. It shows how good of a job the zlib-ng developers have done that it's so performant right out of the gate.
```